File: String/383_RansomNote.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        cnt1 = collections.Counter(ransomNote)
        cnt2 = collections.Counter(magazine)
        for ch in cnt1:
            if cnt2[ch] < cnt1[ch]:
                return False
        return True


class Solution2:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        cnt1 = Counter(ransomNote)
        cnt2 = Counter(magazine)
        if cnt1 & cnt2 == cnt1:
            return True
        return False

File: Java Solutions/383_RansomNote.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        Map<Character, Integer> count = new HashMap<>();
        for (char ch : magazine.toCharArray()) {
            count.put(ch, count.getOrDefault(ch, 0) + 1);
        }
        for (char ch : ransomNote.toCharArray()) {
            if (count.getOrDefault(ch, 0) <= 0) {
                return false;
            }
            count.put(ch, count.get(ch) - 1);
        }
        return true;
    }
}